### PR TITLE
Implement metrics

### PR DESCRIPTION
**Pull request by roby2014**
# Description

**Check the sample for usage**

TLDR: The goal is to have a simple but extendible core class to register, search and debug metrics, with a compiler option to enable it

## Checklist

- [X] Self-review changes.
- [X] Evaluate impact on the documentation.
- [x] Ensure test coverage.
- [X] Write new samples.
- [x] Add entry to the changelog's unreleased section.

### CODE REVIEW

**Comment by RiscadoA** [comment]
Since this PR is still blocked on the other one, I'm marking it as a draft. Does it really depend on the other PR though? They seem mostly unrelated.

**Comment by roby2014** [comment]
It does not depend anymore yeah, you're right.

File: CHANGELOG.md
Comment (RiscadoA) [inline CHANGELOG.md:13]: This shouldn't be here, right?

File: core/include/cubos/core/metrics.hpp
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:25]: Missing CUBOS_CORE_API

File: core/include/cubos/core/metrics.hpp
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:57]: ScopeProfiler perhaps would be a better name? TimeTracker seems unrelated to the metrics system

File: core/include/cubos/core/metrics.hpp
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:74]: Missing @ ingroup core

File: core/include/cubos/core/metrics.hpp
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:78]: Missing in group

File: core/samples/metrics/main.cpp
Comment (RiscadoA) [inline core/samples/metrics/main.cpp:5]: I dont think you need those includes

File: core/src/metrics.cpp
Comment (RiscadoA) [inline core/src/metrics.cpp:47]: Why not store a map from metric names to vectors in the singleton?
This way you wouldnt need this expensive operation here

File: core/include/cubos/core/metrics.hpp
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:79]: Probably a good idea to also add the file and line of code.

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Thanks for working on this :)

File: core/samples/metrics/main.cpp
Comment (roby2014) [reply core/samples/metrics/main.cpp:5]: I do: `std::this_thread::sleep_for(std::chrono::seconds(2));`

Comment (roby2014) [reply core/samples/metrics/main.cpp:5]: at least, the `thread` one, `chrono` can be removed

Comment (RiscadoA) [reply core/samples/metrics/main.cpp]: Ah right, mb

File: core/samples/metrics/main.cpp
Comment (RiscadoA) [inline core/samples/metrics/main.cpp:72]: Won't this fail now, due to __FILE__ + __LINE__? Maybe you can get the name through the `readName` function.

File: core/src/metrics.cpp
Comment (RiscadoA) [inline core/src/metrics.cpp:126]: This is a problem, as the map doesn't have a fixed order. It sorts its keys by lexicographical roder.
Thus, you could end up missing new names and getting old ones repeated when calling this function.
Instead of doing this, keep a separate `vector` for the known metric names.
Each time you add a metric in `::metric` with a name which isn't on the map, you push it to the vector.
Then, here you just access that vector.

**Review by RiscadoA — CHANGES_REQUESTED** [review]

File: core/src/metrics.cpp
Comment (RiscadoA) [inline core/src/metrics.cpp:7]:
```suggestion
#include <cubos/core/reflection/external/primitives.hpp>
#include <cubos/core/reflection/external/string.hpp>
```

File: core/include/cubos/core/metrics.hpp
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:24]: I don't think this matters for generic metrics. Just the name should be enough.
On the profiler side, you can add this information to the name of the metric - this is necessary as to not have conflicts between functions with the same name.

File: core/include/cubos/core/metrics.hpp
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:21]: At this point you don't need to store this anymore, as the values are stored in a map with names as keys anyways.
Thus, you can just keep a `std::deque` of doubles, which should be faster. Which means that, maybe there's no need at all to have this type.

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Sorry for being so nitpicky with these reviews! Since this is something we'll start using everywhere I really want to get the details right. Ty for the all the work :pray:

File: core/include/cubos/core/metrics.hpp
Comment (roby2014) [reply core/include/cubos/core/metrics.hpp:24]: Hmmm, like `myFunction:54:main.cpp` ? The problem will be reading the metric...

Comment (RiscadoA) [reply core/include/cubos/core/metrics.hpp:24]: You would never want to do that though, you would find its name through `readName`.
Btw, we need a way to distinguish profiling metrics from other metrics, maybe we should prefix their name?

Comment (roby2014) [reply core/include/cubos/core/metrics.hpp:24]: `prof__myFunction` and `metric__a` looks good?

Comment (roby2014) [reply core/include/cubos/core/metrics.hpp:24]: IDK, thinking about it later on the UI, not sure how it should look.

Comment (RiscadoA) [reply core/include/cubos/core/metrics.hpp:24]: On the UI we would probably filter out those and give them special treatment. Maybe we should some special character which doesn't appear in function names? For example, `profiling:<file>:<function name:/line number`

Comment (roby2014) [reply core/include/cubos/core/metrics.hpp:24]: so for normal metrics, no prefix, correct?

**Review by RiscadoA — APPROVED** [review]
LGTM! :ship:

File: core/include/cubos/core/metrics.hpp
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:7]:
```suggestion
```

Just noticed this, do not include this in headers! It eats up compile times because it includes pretty much the whole STL

File: core/include/cubos/core/metrics.hpp
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:9]:
```suggestion
```
You only use this in the source file, move it there

**Comment by roby2014** [comment]
Eh no idea why static analysis is failing

**Comment by RiscadoA** [comment]
> Eh no idea why static analysis is failing

It's an error in the action. It's trying to dismiss a review which no longer exists and crashes.
TL;DR: not a problem in your code. @joaomanita @diogomsmiranda pls review when you have the chance

File: core/CMakeLists.txt
Comment (diogomsmiranda) [inline core/CMakeLists.txt:21]: why was 1024 chosen as the max number of entries? Is this arbitrary?

**Review by diogomsmiranda — APPROVED** [review]

**Review by joaomanita — APPROVED** [review]

**Comment by RiscadoA** [comment]
> I also didn't understand the need to make such a sample basically copying from the test. Shouldn't we make something more in tune with what would be a real example?

Agree, we should avoid this, could you improve it @roby2014?

File: core/CMakeLists.txt
Comment (roby2014) [reply core/CMakeLists.txt:21]: Just a dummy value. It can be changed easily through the method `setMaxEntries`

**Comment by diogomsmiranda** [comment]
I would rather have the metrics to be explicitly generated with different names and contexts. But if @RiscadoA is ok with a sample like this It's not a deal breaker for me either.

**Comment by roby2014** [comment]
> I would rather have the metrics to be explicitly generated with different names and contexts. But if @RiscadoA is ok with a sample like this It's not a deal breaker for me either.

Can you give me an example? What should I sample?

**Comment by diogomsmiranda** [comment]
> I also didn't understand the need to make such a sample basically copying from the test. Shouldn't we make something more in tune with what would be a real example?

@roby2014 thank you for addressing this! LGTM

**Comment by RiscadoA** [comment]
You can ignore the failing lint check btw

**Pull request merged**
